File: java/src/test/java/io/github/piyushmishra00/kryon/ConformanceTest.java
package io.github.piyushmishra00.kryon;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;

/**
 * Runs the shared conformance corpus against the Java SDK.
 *
 * <p>The corpus lives at {@code tests/conformance/cases.json} in the repository root and is
 * language-neutral. This class is the Java <em>runner</em>: it maps each case onto this SDK's API
 * and asserts the expectations. Every SDK writes one of these; the corpus itself is never forked.
 */
class ConformanceTest {

    private static final boolean WINDOWS =
            System.getProperty("os.name", "").toLowerCase(Locale.ROOT).contains("win");

    private static final Path CORPUS = Path.of(System.getProperty("kryon.corpus"));
    private static final String HELPER_CLASSES = System.getProperty("kryon.helper.classes");
    private static final String JAVA =
            Path.of(System.getProperty("java.home"), "bin", WINDOWS ? "java.exe" : "java").toString();

    private static final JsonObject CORPUS_JSON = readCorpus();
    private static final List<JsonObject> CASES = readCases();

    private static final Map<String, Class<? extends KryonException>> ERRORS =
            Map.of(
                    "CommandNotFound", CommandNotFoundException.class,
                    "PermissionDenied", PermissionDeniedException.class,
                    "ProcessStartFailed", ProcessStartFailedException.class,
                    "InvalidArguments", InvalidArgumentsException.class,
                    "ProcessFailed", ProcessFailedException.class,
                    "ProcessTimeout", ProcessTimeoutException.class,
                    "ProcessCancelled", ProcessCancelledException.class,
                    "ResourceLimitExceeded", ResourceLimitExceededException.class);

    private static Path scratch;

    private static JsonObject readCorpus() {
        try {
            return JsonParser.parseString(Files.readString(CORPUS, StandardCharsets.UTF_8))
                    .getAsJsonObject();
        } catch (IOException error) {
            throw new UncheckedIOException(error);
        }
    }

    private static List<JsonObject> readCases() {
        List<JsonObject> cases = new ArrayList<>();
        for (JsonElement element : CORPUS_JSON.getAsJsonArray("cases")) {
            cases.add(element.getAsJsonObject());
        }
        return cases;
    }

    private static Path scratch() {
        if (scratch == null) {
            try {
                scratch = Files.createTempDirectory("kryon-conformance-").toRealPath();
            } catch (IOException error) {
                throw new UncheckedIOException(error);
            }
        }
        return scratch;
    }

    // ------------------------------------------------------------------ mapping

    private static String substitute(String value) {
        return value.replace("${TMPDIR}", scratch().toString());
    }

    /**
     * Why a case cannot run here, or null if it can.
     *
     * <p>The JVM cannot set its own environment at runtime, so a case with {@code setup_env} needs
     * that variable supplied by whatever launched the tests. Per {@code spec/conformance.md}, an
     * SDK that cannot satisfy a case reports it as skipped with a reason rather than quietly
     * dropping it.
     */
    private static String skipReason(JsonObject kase) {
        if (kase.has("platforms")) {
            Set<String> platforms = new HashSet<>();
            for (JsonElement element : kase.getAsJsonArray("platforms")) {
                platforms.add(element.getAsString());
            }
            if (!platforms.contains(WINDOWS ? "windows" : "posix")) {
                return "not applicable to this platform";
            }
        }
        if (kase.has("setup_env")) {
            for (var entry : kase.getAsJsonObject("setup_env").entrySet()) {
                String wanted = entry.getValue().getAsString();
                if (!wanted.equals(System.getenv(entry.getKey()))) {
                    return "needs "
                            + entry.getKey()
                            + "="
                            + wanted
                            + " in the environment of the test runner itself; the JVM cannot set "
                            + "its own environment. Run with it set (CI does).";
                }
            }
        }
        return null;
    }

    /**
     * Translates corpus options into this SDK's spelling.
     *
     * <p>The corpus stores durations in seconds; the JVM idiom is {@link Duration}. The semantics
     * are identical, and each SDK converts in its own runner.
     */
    private static ExecutionOptions options(JsonObject kase) {
        JsonObject source =
                kase.has("options") ? kase.getAsJsonObject("options") : new JsonObject();
        ExecutionOptions.Builder builder = ExecutionOptions.builder();

        if (source.has("cwd")) {
            builder.cwd(Path.of(substitute(source.get("cwd").getAsString())));
        }
        if (source.has("env")) {
            for (var entry : source.getAsJsonObject("env").entrySet()) {
                builder.env(
                        entry.getKey(),
                        entry.getValue().isJsonNull() ? null : entry.getValue().getAsString());
            }
        }
        if (source.has("clear_env")) {
            builder.clearEnv(source.get("clear_env").getAsBoolean());
        }
        if (source.has("timeout")) {
            builder.timeout(seconds(source.get("timeout").getAsDouble()));
        }
        if (source.has("kill_grace")) {
            builder.killGrace(seconds(source.get("kill_grace").getAsDouble()));
        }
        if (source.has("max_output_bytes")) {
            builder.maxOutputBytes(source.get("max_output_bytes").getAsLong());
        }
        if (source.has("check")) {
            builder.check(source.get("check").getAsBoolean());
        }
        if (source.has("encoding")) {
            builder.charset(StandardCharsets.UTF_8);
        }
        if (kase.has("stdin")) {
            builder.stdin(kase.get("stdin").getAsString());
        }
        return builder.build();
    }

    private static Duration seconds(double value) {
        return Duration.ofNanos(Math.round(value * 1_000_000_000L));
    }

    private record Command(String executable, List<String> arguments, boolean shell) {}

    private static Command command(JsonObject kase) {
        if (kase.has("shell_command")) {
            return new Command(
                    kase.getAsJsonObject("shell_command")
                            .get(WINDOWS ? "windows" : "posix")
                            .getAsString(),
                    List.of(),
                    true);
        }
        List<String> args = new ArrayList<>();
        if (kase.has("args")) {
            for (JsonElement element : kase.getAsJsonArray("args")) {
                args.add(substitute(element.getAsString()));
            }
        }
        if (kase.has("executable")) {
            return new Command(kase.get("executable").getAsString(), args, false);
        }

        List<String> argv = new ArrayList<>();
        argv.add("-cp");
        argv.add(HELPER_CLASSES);
        argv.add(ConformanceHelper.class.getName());
        argv.addAll(args);
        return new Command(JAVA, argv, false);
    }

    // --------------------------------------------------------------- assertions

    private static void assertResult(JsonObject kase, ExecutionResult result) {
        JsonObject e = kase.getAsJsonObject("expect");
        String id = kase.get("id").getAsString();

        if (e.has("exit_code")) {
            assertEquals(e.get("exit_code").getAsInt(), result.exitCode().orElse(Integer.MIN_VALUE), id);
        }
        if (e.has("termination")) {
            assertEquals(e.get("termination").getAsString(), result.termination().name(), id);
        }
        if (e.has("ok")) {
            assertEquals(e.get("ok").getAsBoolean(), result.ok(), id);
        }
        if (e.has("stdout")) {
            assertEquals(e.get("stdout").getAsString(), result.stdout(), id);
        }
        if (e.has("stderr")) {
            assertEquals(e.get("stderr").getAsString(), result.stderr(), id);
        }
        if (e.has("stdout_contains")) {
            assertTrue(result.stdout().contains(e.get("stdout_contains").getAsString()), id);
        }
        if (e.has("stderr_contains")) {
            assertTrue(result.stderr().contains(e.get("stderr_contains").getAsString()), id);
        }
        if (e.has("stdout_truncated")) {
            assertEquals(e.get("stdout_truncated").getAsBoolean(), result.stdoutTruncated(), id);
        }
        if (e.has("stdout_bytes_at_most")) {
            assertTrue(
                    result.stdoutBytes().length <= e.get("stdout_bytes_at_most").getAsLong(),
                    id + ": " + result.stdoutBytes().length + " bytes");
        }
        if (e.has("duration_at_most")) {
            long limit = Math.round(e.get("duration_at_most").getAsDouble() * 1000);
            assertTrue(result.duration().toMillis() <= limit, id + ": " + result.duration());
        }
        if (e.has("duration_at_least")) {
            long limit = Math.round(e.get("duration_at_least").getAsDouble() * 1000);
            assertTrue(result.duration().toMillis() >= limit, id + ": " + result.duration());
        }
        if (e.has("signal_present")) {
            assertEquals(e.get("signal_present").getAsBoolean(), result.signal().isPresent(), id);
        }
        if (e.has("stdout_is_bytes") && e.get("stdout_is_bytes").getAsBoolean()) {
            // Java always exposes both views; the byte view is the primitive one, and the
            // corpus intent -- "no codec was guessed" -- is satisfied by it existing.
            assertTrue(result.stdoutBytes().length > 0, id);
        }
        if (e.has("stdout_contains_bytes")) {
            byte[] needle = hex(e.get("stdout_contains_bytes").getAsString());
            assertTrue(indexOf(result.stdoutBytes(), needle) >= 0, id);
        }
        if (e.has("stdout_is_dir")) {
            try {
                Path expected = Path.of(substitute(e.get("stdout_is_dir").getAsString())).toRealPath();
                Path actual = Path.of(result.stdout().trim()).toRealPath();
                assertEquals(expected, actual, id);
            } catch (IOException error) {
                throw new UncheckedIOException(error);
            }
        }
    }

    private static byte[] hex(String value) {
        byte[] bytes = new byte[value.length() / 2];
        for (int i = 0; i < bytes.length; i++) {
            bytes[i] = (byte) Integer.parseInt(value.substring(i * 2, i * 2 + 2), 16);
        }
        return bytes;
    }

    private static int indexOf(byte[] haystack, byte[] needle) {
        outer:
        for (int i = 0; i + needle.length <= haystack.length; i++) {
            for (int j = 0; j < needle.length; j++) {
                if (haystack[i + j] != needle[j]) {
                    continue outer;
                }
            }
            return i;
        }
        return -1;
    }

    // ------------------------------------------------------------------- tests

    @TestFactory
    Stream<DynamicTest> execute() {
        return CASES.stream()
                .filter(c -> !"spawn".equals(c.get("api").getAsString()))
                .map(
                        kase ->
                                DynamicTest.dynamicTest(
                                        kase.get("id").getAsString(),
                                        () -> {
                                            String skip = skipReason(kase);
                                            Assumptions.assumeTrue(skip == null, skip);

                                            Runtime runtime = new Runtime();
                                            Command cmd = command(kase);
                                            ExecutionOptions opts = options(kase);
                                            JsonObject expect = kase.getAsJsonObject("expect");

                                            if (expect.has("raises")) {
                                                assertThrows(
                                                        ERRORS.get(expect.get("raises").getAsString()),
                                                        () -> invoke(runtime, cmd, opts),
                                                        kase.get("id").getAsString());
                                                return;
                                            }
                                            assertResult(kase, invoke(runtime, cmd, opts));
                                        }));
    }

    private static ExecutionResult invoke(Runtime runtime, Command cmd, ExecutionOptions opts) {
        return cmd.shell()
                ? runtime.executeShell(cmd.executable(), opts)
                : runtime.execute(cmd.executable(), cmd.arguments(), opts);
    }

    @TestFactory
    Stream<DynamicTest> spawn() {
        return CASES.stream()
                .filter(c -> "spawn".equals(c.get("api").getAsString()))
                .map(
                        kase ->
                                DynamicTest.dynamicTest(
                                        kase.get("id").getAsString(),
                                        () -> {
                                            String skip = skipReason(kase);
                                            Assumptions.assumeTrue(skip == null, skip);
                                            runSpawnCase(kase);
                                        }));
    }

    private static void runSpawnCase(JsonObject kase) throws Exception {
        Runtime runtime = new Runtime();
        Command cmd = command(kase);
        JsonObject e = kase.getAsJsonObject("expect");
        String id = kase.get("id").getAsString();

        if (kase.has("scope_exit_only")) {
            // Deliberately not try-with-resources: this case is about what leaving the
            // scope does, so the close has to be the observable event under test.
            KryonProcess proc = runtime.spawn(cmd.executable(), cmd.arguments(), options(kase));
            assertTrue(proc.running(), id);
            proc.close();
            assertEquals(e.get("running_after_scope").getAsBoolean(), proc.running(), id);
            return;
        }

        try (KryonProcess proc =
                runtime.spawn(cmd.executable(), cmd.arguments(), options(kase))) {

            if (kase.has("terminate_after")) {
                Thread.sleep(Math.round(kase.get("terminate_after").getAsDouble() * 1000));
                proc.terminate();
                ExecutionResult result = proc.await(Duration.ofSeconds(20));
                assertEquals(e.get("running_after_terminate").getAsBoolean(), proc.running(), id);
                if (e.has("duration_at_most")) {
                    long limit = Math.round(e.get("duration_at_most").getAsDouble() * 1000);
                    assertTrue(result.duration().toMillis() <= limit, id);
                }
                return;
            }

            if (kase.has("write")) {
                for (JsonElement element : kase.getAsJsonArray("write")) {
                    proc.write(element.getAsString());
                }
            }
            if (kase.has("close_stdin") && kase.get("close_stdin").getAsBoolean()) {
                proc.closeStdin();
            }

            List<OutputChunk> chunks = new ArrayList<>();
            for (OutputChunk chunk : proc.output()) {
                chunks.add(chunk);
            }
            ExecutionResult result = proc.await(Duration.ofSeconds(20));

            StringBuilder text = new StringBuilder();
            for (OutputChunk chunk : chunks) {
                text.append(new String(chunk.data(), StandardCharsets.UTF_8));
            }

            if (e.has("streamed_chunks_at_least")) {
                assertTrue(chunks.size() >= e.get("streamed_chunks_at_least").getAsInt(), id);
            }
            if (e.has("stdout_contains")) {
                assertTrue(text.toString().contains(e.get("stdout_contains").getAsString()), id);
            }
            if (e.has("stdout_contains_last")) {
                assertTrue(
                        text.toString().contains(e.get("stdout_contains_last").getAsString()), id);
            }
            if (e.has("exit_code")) {
                assertEquals(
                        e.get("exit_code").getAsInt(), result.exitCode().orElse(Integer.MIN_VALUE), id);
            }
        }
    }

    // -------------------------------------------------------- corpus integrity

    @Test
    void everyCaseHasARunner() {
        Set<String> apis = CASES.stream().map(c -> c.get("api").getAsString()).collect(Collectors.toSet());
        assertTrue(
                Set.of("execute", "execute_shell", "spawn").containsAll(apis),
                "the corpus uses an api this runner does not handle: " + apis);
    }

    @Test
    void idsAreUnique() {
        List<String> ids = CASES.stream().map(c -> c.get("id").getAsString()).toList();
        assertEquals(ids.size(), Set.copyOf(ids).size(), "duplicate case id in the corpus");
    }

    @Test
    void everyCaseExplainsItself() {
        List<String> missing =
                CASES.stream()
                        .filter(c -> !c.has("why"))
                        .map(c -> c.get("id").getAsString())
                        .toList();
        assertTrue(missing.isEmpty(), "cases missing a 'why': " + missing);
    }

    @Test
    void corpusIsNotEmpty() {
        JsonArray cases = CORPUS_JSON.getAsJsonArray("cases");
        assertTrue(cases.size() >= 30, "the corpus looks truncated: " + cases.size() + " cases");
    }
}

File: kotlin/src/test/kotlin/io/github/piyushmishra00/kryon/coroutines/ConformanceTest.kt
package io.github.piyushmishra00.kryon.coroutines

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import java.nio.file.Files
import java.nio.file.Path
import java.util.Locale
import kotlin.math.roundToLong
import kotlin.test.assertContains
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assumptions
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestFactory
import org.junit.jupiter.api.assertThrows

/**
 * Runs the shared conformance corpus against the Kotlin SDK.
 *
 * The corpus lives at `tests/conformance/cases.json` in the repository root and is
 * language-neutral. This class is the Kotlin *runner*: it maps each case onto this SDK's API and
 * asserts the expectations. Every SDK writes one of these; the corpus itself is never forked.
 */
class ConformanceTest {

    private companion object {
        val windows: Boolean =
            System.getProperty("os.name", "").lowercase(Locale.ROOT).contains("win")

        val javaExe: String =
            Path.of(System.getProperty("java.home"), "bin", if (windows) "java.exe" else "java")
                .toString()

        val classpath: String = System.getProperty("kryon.helper.classpath")

        val corpus: JsonObject =
            JsonParser.parseString(
                Files.readString(Path.of(System.getProperty("kryon.corpus"))),
            ).asJsonObject

        val cases: List<JsonObject> = corpus.getAsJsonArray("cases").map { it.asJsonObject }

        val scratch: Path by lazy {
            Files.createTempDirectory("kryon-conformance-").toRealPath()
        }

        val errors: Map<String, Class<out KryonException>> = mapOf(
            "CommandNotFound" to CommandNotFoundException::class.java,
            "PermissionDenied" to PermissionDeniedException::class.java,
            "ProcessStartFailed" to ProcessStartFailedException::class.java,
            "InvalidArguments" to InvalidArgumentsException::class.java,
            "ProcessFailed" to ProcessFailedException::class.java,
            "ProcessTimeout" to ProcessTimeoutException::class.java,
            "ProcessCancelled" to ProcessCancelledException::class.java,
            "ResourceLimitExceeded" to ResourceLimitExceededException::class.java,
        )
    }

    private fun substitute(value: String) = value.replace("\${TMPDIR}", scratch.toString())

    /**
     * Why a case cannot run here, or null if it can.
     *
     * The JVM cannot set its own environment at runtime, so a case with `setup_env` needs that
     * variable supplied by whatever launched the tests. Per `spec/conformance.md`, an SDK that
     * cannot satisfy a case reports it as skipped with a reason rather than quietly dropping it.
     */
    private fun skipReason(kase: JsonObject): String? {
        kase.getAsJsonArray("platforms")?.let { platforms ->
            val wanted = if (windows) "windows" else "posix"
            if (platforms.none { it.asString == wanted }) return "not applicable to this platform"
        }
        kase.getAsJsonObject("setup_env")?.entrySet()?.forEach { (key, value) ->
            if (System.getenv(key) != value.asString) {
                return "needs $key=${value.asString} in the environment of the test runner " +
                    "itself; the JVM cannot set its own environment. Run with it set (CI does)."
            }
        }
        return null
    }

    /**
     * Translates corpus options into this SDK's spelling.
     *
     * The corpus stores durations in seconds; Kotlin's idiom is [Duration]. The semantics are
     * identical, and each SDK converts in its own runner.
     */
    private fun options(kase: JsonObject): ExecutionOptions {
        val source = kase.getAsJsonObject("options") ?: JsonObject()

        fun duration(key: String): Duration? =
            source.get(key)?.asDouble?.let { (it * 1000).roundToLong().milliseconds }

        val env = source.getAsJsonObject("env")?.entrySet()?.associate { (key, value) ->
            key to if (value.isJsonNull) null else value.asString
        } ?: emptyMap()

        return ExecutionOptions(
            cwd = source.get("cwd")?.asString?.let { Path.of(substitute(it)) },
            env = env,
            clearEnv = source.get("clear_env")?.asBoolean,
            stdin = kase.get("stdin")?.asString?.toByteArray(),
            timeout = duration("timeout"),
            killGrace = duration("kill_grace"),
            maxOutputBytes = source.get("max_output_bytes")?.asLong,
            charset = source.get("encoding")?.let { Charsets.UTF_8 },
            check = source.get("check")?.asBoolean,
        )
    }

    private data class Command(
        val executable: String,
        val arguments: List<String>,
        val shell: Boolean,
    )

    private fun command(kase: JsonObject): Command {
        kase.getAsJsonObject("shell_command")?.let {
            return Command(it.get(if (windows) "windows" else "posix").asString, emptyList(), true)
        }
        val args = kase.getAsJsonArray("args")?.map { substitute(it.asString) } ?: emptyList()
        kase.get("executable")?.let { return Command(it.asString, args, false) }

        return Command(
            javaExe,
            listOf("-cp", classpath, ConformanceHelper::class.java.name) + args,
            false,
        )
    }

    private suspend fun invoke(cmd: Command, opts: ExecutionOptions): ExecutionResult {
        val runtime = Runtime()
        return if (cmd.shell) {
            runtime.executeShell(cmd.executable, opts)
        } else {
            runtime.execute(cmd.executable, cmd.arguments, opts)
        }
    }

    private fun assertResult(kase: JsonObject, result: ExecutionResult) {
        val e = kase.getAsJsonObject("expect")
        val id = kase.get("id").asString

        e.get("exit_code")?.let { assertEquals(it.asInt, result.exitCode, id) }
        e.get("termination")?.let { assertEquals(it.asString, result.termination.name, id) }
        e.get("ok")?.let { assertEquals(it.asBoolean, result.ok, id) }
        e.get("stdout")?.let { assertEquals(it.asString, result.stdout, id) }
        e.get("stderr")?.let { assertEquals(it.asString, result.stderr, id) }
        e.get("stdout_contains")?.let { assertContains(result.stdout, it.asString, message = id) }
        e.get("stderr_contains")?.let { assertContains(result.stderr, it.asString, message = id) }
        e.get("stdout_truncated")?.let { assertEquals(it.asBoolean, result.stdoutTruncated, id) }
        e.get("stdout_bytes_at_most")?.let {
            assertTrue(result.stdoutRaw.size <= it.asLong, "$id: ${result.stdoutRaw.size} bytes")
        }
        e.get("duration_at_most")?.let {
            assertTrue(
                result.duration <= (it.asDouble * 1000).roundToLong().milliseconds,
                "$id: ${result.duration}",
            )
        }
        e.get("duration_at_least")?.let {
            assertTrue(
                result.duration >= (it.asDouble * 1000).roundToLong().milliseconds,
                "$id: ${result.duration}",
            )
        }
        e.get("signal_present")?.let { assertEquals(it.asBoolean, result.signal != null, id) }
        e.get("stdout_is_bytes")?.let {
            if (it.asBoolean) assertTrue(result.stdoutRaw.isNotEmpty(), id)
        }
        e.get("stdout_contains_bytes")?.let {
            val needle = it.asString.chunked(2).map { pair -> pair.toInt(16).toByte() }
                .toByteArray()
            assertTrue(indexOf(result.stdoutRaw, needle) >= 0, id)
        }
        e.get("stdout_is_dir")?.let {
            val expected = Path.of(substitute(it.asString)).toRealPath()
            val actual = Path.of(result.stdout.trim()).toRealPath()
            assertEquals(expected, actual, id)
        }
    }

    private fun indexOf(haystack: ByteArray, needle: ByteArray): Int {
        outer@ for (i in 0..haystack.size - needle.size) {
            for (j in needle.indices) {
                if (haystack[i + j] != needle[j]) continue@outer
            }
            return i
        }
        return -1
    }

    @TestFactory
    fun execute(): List<DynamicTest> =
        cases.filter { it.get("api").asString != "spawn" }.map { kase ->
            DynamicTest.dynamicTest(kase.get("id").asString) {
                val skip = skipReason(kase)
                Assumptions.assumeTrue(skip == null, skip)

                runBlocking {
                    val cmd = command(kase)
                    val opts = options(kase)
                    val expect = kase.getAsJsonObject("expect")

                    val raises = expect.get("raises")?.asString
                    if (raises != null) {
                        val thrown = assertThrows<KryonException> {
                            runBlocking { invoke(cmd, opts) }
                        }
                        assertEquals(errors[raises], thrown.javaClass, kase.get("id").asString)
                        return@runBlocking
                    }
                    assertResult(kase, invoke(cmd, opts))
                }
            }
        }

    @TestFactory
    fun spawn(): List<DynamicTest> =
        cases.filter { it.get("api").asString == "spawn" }.map { kase ->
            DynamicTest.dynamicTest(kase.get("id").asString) {
                val skip = skipReason(kase)
                Assumptions.assumeTrue(skip == null, skip)
                runBlocking { runSpawnCase(kase) }
            }
        }

    private suspend fun runSpawnCase(kase: JsonObject) {
        val runtime = Runtime()
        val cmd = command(kase)
        val e = kase.getAsJsonObject("expect")
        val id = kase.get("id").asString

        val proc = runtime.spawn(cmd.executable, cmd.arguments, options(kase))

        if (kase.get("scope_exit_only")?.asBoolean == true) {
            assertTrue(proc.running, id)
            proc.closeAndJoin()
            assertEquals(e.get("running_after_scope").asBoolean, proc.running, id)
            return
        }

        proc.use { process ->
            kase.get("terminate_after")?.let { after ->
                delay((after.asDouble * 1000).roundToLong())
                process.terminate()
                val result = process.await(20.seconds)
                assertEquals(e.get("running_after_terminate").asBoolean, process.running, id)
                e.get("duration_at_most")?.let {
                    assertTrue(
                        result.duration <= (it.asDouble * 1000).roundToLong().milliseconds,
                        id,
                    )
                }
                return@use
            }

            kase.getAsJsonArray("write")?.forEach { process.write(it.asString) }
            if (kase.get("close_stdin")?.asBoolean == true) process.closeStdin()

            val chunks = process.output.toList()
            val result = process.await(20.seconds)
            val text = chunks.fold(ByteArray(0)) { acc, chunk -> acc + chunk.data }
                .toString(Charsets.UTF_8)

            e.get("streamed_chunks_at_least")?.let {
                assertTrue(chunks.size >= it.asInt, id)
            }
            e.get("stdout_contains")?.let { assertContains(text, it.asString, message = id) }
            e.get("stdout_contains_last")?.let { assertContains(text, it.asString, message = id) }
            e.get("exit_code")?.let { assertEquals(it.asInt, result.exitCode, id) }
        }
    }

    @Test
    fun idsAreUnique() {
        val ids = cases.map { it.get("id").asString }
        assertEquals(ids.size, ids.toSet().size, "duplicate case id in the corpus")
    }

    @Test
    fun everyCaseExplainsItself() {
        val missing = cases.filter { !it.has("why") }.map { it.get("id").asString }
        assertTrue(missing.isEmpty(), "cases missing a 'why': $missing")
    }

    @Test
    fun everyCaseHasARunner() {
        val apis = cases.map { it.get("api").asString }.toSet()
        assertTrue(
            setOf("execute", "execute_shell", "spawn").containsAll(apis),
            "the corpus uses an api this runner does not handle: $apis",
        )
    }
}
